### PR TITLE
Experimenting with Jest's snapshot feature

### PR DESCRIPTION
**Pull request by jhilde**
The bug I introduced described in #1192 has been annoying me. Essentially, I added a mandatory prop to a component. I didn't realize it was being used by another page, so I didn't change it. Tests were clear because that page isn't well tested. 

As Kevin pointed out, perhaps Flow or Typescript would alleviate my craving of some sort of type safety that's compiler enforced. _We won't be using Flow or Typescript anytime soon._

So I started to think about how we can ensure that each of the React components are tested. 

I'm thinking using Jest snapshots for very basic testing of React components could go a long way. Essentially, Jest takes a snapshot the first time the tests are run. Those snapshots are committed alongside jest test js.

So I wanted to test a couple of components. Two with existing tests `SectionPage` and `FlexibleRoster` and one without `SectionHeader`.

Would love your thoughts @kevinrobinson and @alexsoble. Note I had to downgrade react-test-renderer to the version of react we're using. Doesn't look like that was used anywhere else.

### CODE REVIEW

**Comment by kevinrobinson** [comment]
@jhilde This seems great to explore!  👍 

I'm not sure how this improves the case you ran into though, as compared to just adding smoke tests like [this](https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/src/App.test.js#L5).  Missing props is more about just having a test at all rather than depth of coverage, so adding one-line tests seem the best immediate bang for the buck.

One other downside for this might be fatigue around snapshots changing on every change, or that we're not actually good at catching bugs by reviewing snapshot files.  That being said, I'd vote to just try it out and see how we all like it.

File: spec/javascripts/section/section_page_spec.jsx [hunk 1]
```diff
@@ -80,4 +81,28 @@ describe('SectionPage', function() {
       expect(firstDataRows[0].innerHTML).toEqual('<a href="/students/2">Duck, Donald</a>');
     });
   });
+
+  it('renders correctly', () => {
+    const props = {
```
Comment (kevinrobinson) [inline spec/javascripts/section/section_page_spec.jsx]: One pattern I like in test files is:

```
function testProps(props) {
  return {
    ...props,
     foo: 'default'
  };
}
...
const props = testProps({ bar: 'cfdsya' });
````

Comment (kevinrobinson) [reply spec/javascripts/section/section_page_spec.jsx]: (as a way to modify default fixture values so you don't have to duplicate them). 
 There's also a `fixtures` folder or file for common ones (like a default `currentEducator`).

**Comment by kevinrobinson** [comment]
Also, to clarify https://github.com/studentinsights/studentinsights/issues/1192#issuecomment-339755769 I think it'd be totally reasonable to experiment with weak mode Flow and see how we all find it (similar to what you're proposing here).

**Comment by jhilde** [comment]
@kevinrobinson Agree that a simple test can do the job. The challenge is more in creating props (providing fixtures for common things will work well). 

In looking at our tests, the majority are asking, does this look right. Some count the number of columns, some search for text, etc. It seems like automated snapshots would actually do most of that testing, with less code, and more completely. Agreed, that we update a core React Component, several snapshots could change. But with auto recreation of snapshots, we can then look at the diffs to make sure the output changed in the expected way.

Mostly it was just a, I have a few minutes, what should I play with. And also excited we're on more modern tooling now. ;-)

**Comment by kevinrobinson** [comment]
👍 For adding smoke tests, here are some ways fixtures are used in other places in the project: [local to the test](https://github.com/studentinsights/studentinsights/blob/adedc3bc701fdf926d45c3962e5ea44505463da5/spec/javascripts/student_profile/services_list_spec.jsx#L23), [shared across a product feature](https://github.com/studentinsights/studentinsights/blob/master/spec/javascripts/student_profile/fixtures.jsx) and [shared across the app](https://github.com/studentinsights/studentinsights/tree/master/spec/javascripts/fixtures).  If you need to make new ones that are complex, grabbing data from the console in development mode is a good way to re-use the Ruby demo data generators.

For snapshots, let's just try it!  It'll definitely be more complete in some sense, but it relies on us humans to look at those snapshots and diffs to know if the feature is working correctly. :)  In this PR here, does the roster choose the initial sort column correctly?  For me, it's harder to see that in the snapshot than in:

```
expect($(el).find('th:first').hasClass('sort-down')).toEqual(true);
```

and not much more pain to write.  For other components for things like visualizations, where TDD makes no sense and you want to verify you don't regress visual layout or styling, snapshots seem like the perfect test API.  I'm optimistic that snapshots would be helpful enough to be worth trying out!

But if we want to improve catching errors for missing props, we should just do the plain boring thing and add some smoke tests to components that don't have them :)

**Comment by kevinrobinson** [comment]
Moving over from email.  I'd vote to experiment and try using snapshots like you're doing here, I think I've said that in all of my comments :)

> It seems like automated snapshots would actually do most of that testing, with less code, and more completely

I agree with this!  Certainly if we added snapshots for all components, those tests would be quite complete and fail anytime the product changes.  I'm not sure that this helps our users, or even us. :)

The cost that snapshots introduces is higher false-positive test failures, since snapshot tests break on any product change.  And that means test fatigue and friction for every developer on the team for every product improvement they make.  In a weird way, that approach would encourage us to make refactoring improvements that were invisible to users, but discourage us from making product improvements.  I've certainly worked on systems where tests functioned this way.  If you have other perspectives you can share that'd be helpful for me in understanding!

To re-iterate, I'd vote to experiment and try using snapshots like you're doing here, and to experiment with them on new things we build.  This seems great!  And I'd be excited about pull requests that added smoke tests to all components proactively.  But I'd vote against spending time to add snapshot tests to every component pro-actively.

**Comment by jhilde** [comment]
I'm going to close this without merging...and perhaps do a bit more with it. This was originally meant as an experiment.

**Pull request closed without merging**